packages/make-styles/src/makeStyles.ts

```diff
@@ -36,41 +37,55 @@ export function makeStyles<Selectors, Tokens>(definitions: MakeStylesDefinition<
     const overrides: MakeStylesMatchedDefinitions = {};
     let overridesCx: string = '';
 
-    classNames.forEach(className => {
-      if (typeof className === 'string') {
-        if (className === '') {
-          return;
-        }
+    // arguments are parsed intentionally manually to avoid double loops as TS transforms rest via an additional loop
+    for (let i = 2; i < arguments.length; i++) {
+      // eslint-disable-next-line prefer-rest-params
+      const classNames = arguments[i];
+
+      if (typeof classNames === 'string' && classNames !== '') {
+        // .split() is an expensive call, it's faster to ensure that string contains any spaces before splitting
+        if (classNames.indexOf(' ') === -1) {
+          const definition = DEFINITION_LOOKUP_TABLE[classNames];
 
-        className.split(' ').forEach(cName => {
-          if (DEFINITION_LOOKUP_TABLE[cName] !== undefined) {
-            overrides[DEFINITION_LOOKUP_TABLE[cName][0]] = DEFINITION_LOOKUP_TABLE[cName][1];
-            overridesCx += cName;
+          if (definition !== undefined) {
+            overrides[definition[0]] = DEFINITION_LOOKUP_TABLE[classNames][1];
+            overridesCx += classNames;
           } else {
-            nonMakeClasses += cName + ' ';
+            nonMakeClasses += classNames + ' ';
           }
-        });
+        } else {
+          classNames.split(' ').forEach(className => {
+            const definition = DEFINITION_LOOKUP_TABLE[className];
+
+            if (definition !== undefined) {
+              overrides[definition[0]] = definition[1];
+              overridesCx += className;
+            } else {
+              nonMakeClasses += className + ' ';
+            }
+          });
+        }
       }
-    });
+    }
 
     let matchedIndexes = '';
+    const matchedDefinitions: MakeStylesMatchedDefinitions[] = [];
 
-    const matchedDefinitions = resolvedDefinitions.reduce<MakeStylesMatchedDefinitions[]>((acc, definition, i) => {
-      const matcherFn = definition[0];
+    for (let i = 0, l = resolvedDefinitions.length; i < l; i++) {
```

Comment:
I have an idea, let me know what you think. 
Instead of iterating `resolvedDefinitions` here for building the cache key, how about we use `selectors` for building the cache key:
```
    for (const k in selectors) {
      const s = selectors[k];
      matchedIndexes += s === true ? '1' : s === false ? '0' : 'u';
    }
```
In this case, we would want to limit `selectors` to be `boolean | undefined` value.

Then only when there is a cache miss, we iterate through `resolvedDefinitions`  as is. I think in real world, there are going to be :
1. more cache hit than miss 
2. more entries in `resolvedDefinitions` than `selectors`